### PR TITLE
Only show wikis tab when there are wikis

### PR DESCRIPTION
**Pull request by NobodysNightmare**
Effectively we only want to allow disabling the internal wiki for now, but we'll have to answer certain usability questions on what happens when a provider is deleted/disabled anyways, so we'll make it technically possible to disable any provider, even if we might not offer it for all of them out of the box.

The most likely starting scenario to have no wikis tab is that the internal wikis are disabled and no external wiki is configured.

### Ticket
https://community.openproject.org/wp/72969

### CODE REVIEW

File: frontend/src/app/core/config/configuration.service.ts [hunk 1]
```diff
@@ -147,6 +147,10 @@ export class ConfigurationService {
     return moment.localeData(I18n.locale).firstDayOfWeek();
   }
 
+  public get wikisAvailable():boolean {
+    return this.systemPreference('wikisAvailable');
```
Comment (NobodysNightmare) [inline frontend/src/app/core/config/configuration.service.ts:151]: 🟡 **Note for reviewer:** This change is notable, because it's the only change outside the `modules/wiki` folder. Therefore this is strictly speaking a violation of the "module promise". This code would remain, even if the wikis module was removed.

It wouldn't be too breaking of a change right now:

* no callers left for this getter
* the presence of the getter does not break the rest of the service

But it still feels wrong. Do we have an established way to extend existing services via a module?

Comment (mereghost) [reply frontend/src/app/core/config/configuration.service.ts:151]: We could have some sort of pubsub / queue structure that modules could publish to and then the "core" services consume it.

This as a technique used by the Incas way back in the JQuery days.

File: modules/storages/app/common/storages/adapters/registry.rb [hunk 2]
```diff
@@ -45,14 +45,14 @@ class Resolver < Dry::Container::Resolver
         include TaggedLogging
 
         def call(container, key)
-          with_tagged_logger("Adapters::Registry") do
+          with_tagged_logger("Storages::Adapters::Registry") do
```
Comment (mereghost) [inline modules/storages/app/common/storages/adapters/registry.rb:48]: 🟢 Oh! Good catch. Another option is to remove the argument and it should log as `Storages::Adapters::Registry::Resolver`.

No real preference here.

File: modules/wikis/app/models/wikis/provider.rb [hunk 3]
```diff
@@ -33,5 +33,15 @@ class Provider < ApplicationRecord
     self.table_name = "wiki_providers"
 
     has_many :page_links, dependent: :destroy
+
+    scope :enabled, -> { where(enabled: true) }
+
+    class << self
+      def registry_prefix = raise NotImplementedError, "SubclassResponsibility"
```
Comment (mereghost) [inline modules/wikis/app/models/wikis/provider.rb:40]: 🟡 I'd recommend creating (or making the storages exception usable globally) a `SubclassResponsibility` as the semantics of `NotImplementedError` are: 

> Raised when a feature is not implemented on the current platform.

File: modules/wikis/app/models/wikis/internal_provider.rb [hunk 4]
```diff
@@ -30,6 +30,10 @@
 
 module Wikis
   class InternalProvider < Provider
+    class << self
```
Comment (mereghost) [inline modules/wikis/app/models/wikis/internal_provider.rb:33]: 🟢 For ease of use I'd add an instance level reader, as I'm too lazy to type `object.class.method` xD

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/test.rb [hunk 5]
```diff
@@ -23,20 +23,20 @@
...
 #
 # See COPYRIGHT and LICENSE files for more details.
 #++
 
-module Workflows
-  class BlankslateComponent < ApplicationComponent
-    include OpPrimer::ComponentHelpers
+# This query is only intended for demo purposes and can be deleted once we have real queries
+module Wikis::Adapters::Providers::Internal::Queries
+  class Test
```
Comment (mereghost) [inline modules/wikis/app/services/wikis/adapters/providers/internal/queries/test.rb:33]: 🟢 Expanding the module definition helps a lot when referencing other code in the same namespace, by not requiring you to fully qualify the objects.

File: modules/wikis/app/services/wikis/adapters/providers/internal/registry.rb [hunk 6]
```diff
@@ -23,22 +23,37 @@
...
 # See COPYRIGHT and LICENSE files for more details.
 #++
 
-class AddUniquenessForActiveSprints < ActiveRecord::Migration[8.0]
-  disable_ddl_transaction!
+module Wikis
+  module Adapters
+    module Providers
+      module Internal
+        Registry = Dry::Container::Namespace.new("internal") do
```
Comment (mereghost) [inline modules/wikis/app/services/wikis/adapters/providers/internal/registry.rb:35]: <3

**Review by mereghost — APPROVED** [review]
Overall looks great. I just left a bunch of comments that don't required immediate action, just food for discussion and reflection.

File: modules/wikis/app/models/wikis/internal_provider.rb (hunk 4 above)
Comment (NobodysNightmare) [reply modules/wikis/app/models/wikis/internal_provider.rb:33]: To be fair: You usually shouldn't have to call `provider.class.registry_prefix` at all. What you are supposed to call is `provider.resolve("queries.test")`.

This is essentially the evolution of `Registry["#{storage}.queries.test"].new(storage)`, which is the common pattern we have in the storages module. I hope to be able to replace most of those usages with `resolve` and only need "external" access to the prefix in a few edge cases.

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/test.rb (hunk 5 above)
Comment (NobodysNightmare) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/test.rb:33]: That's a good point. I remembered that there was a reason to prefer writing out all modules separately, but I didn't remember.

On the other hand, writing out all the modules nests us 8 characters deeper...

(I'll not change anything on the `Test` query, but we should consider your comment for future queries)

File: modules/wikis/app/models/wikis/provider.rb (hunk 3 above)
Comment (NobodysNightmare) [reply modules/wikis/app/models/wikis/provider.rb:40]: I am working on this, but seeing the scope, I'll probably move this into a separate PR.

File: modules/wikis/app/models/wikis/internal_provider.rb (hunk 4 above)
Comment (mereghost) [reply modules/wikis/app/models/wikis/internal_provider.rb:33]: So you intend to proxy the access to the registry through the provider? Sounds interesting, let's give it a try. =)

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/test.rb (hunk 5 above)
Comment (mereghost) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/test.rb:33]: I feel you, but that's the penalty of deeply nested namespaces. =/
